src/parser/base.rs

```diff
@@ -0,0 +1,239 @@
+use std::{borrow::ToOwned, string::String, vec::Vec};
+
+use crate::field::LurkField;
+use base_x;
+use nom::{
+    branch::alt,
+    bytes::complete::{tag, take_till},
+    character::complete::satisfy,
+    combinator::value,
+    error::context,
+    IResult, InputTakeAtPosition,
+};
+
+use crate::parser::{
+    error::{map_parse_err, ParseError, ParseErrorKind},
+    Span,
+};
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub enum LitBase {
+    Bin,
+    Oct,
+    Dec,
+    Hex,
+}
+
+impl Default for LitBase {
+    fn default() -> Self {
+        Self::Hex
+    }
+}
+
+impl LitBase {
+    pub fn parse_code<F: LurkField>(
+        i: Span<'_>,
+    ) -> IResult<Span<'_>, Self, ParseError<Span<'_>, F>> {
+        alt((
+            value(Self::Bin, tag("b")),
+            value(Self::Oct, tag("o")),
+            value(Self::Dec, tag("d")),
+            value(Self::Hex, tag("x")),
+        ))(i)
+    }
+
+    /// Get the code corresponding to the base algorithm.
+    pub fn code(&self) -> char {
+        match self {
+            Self::Bin => 'b',
+            Self::Oct => 'o',
+            Self::Dec => 'd',
+            Self::Hex => 'x',
+        }
+    }
+
+    pub fn base_digits(&self) -> &str {
+        match self {
+            Self::Bin => "01",
+            Self::Oct => "01234567",
+            Self::Dec => "0123456789",
+            Self::Hex => "0123456789abcdef",
+        }
+    }
+
+    pub fn radix(&self) -> u32 {
+        match self {
+            Self::Bin => 2,
+            Self::Oct => 8,
+            Self::Dec => 10,
+            Self::Hex => 16,
+        }
+    }
+
+    pub fn is_digit(&self, x: char) -> bool {
+        self.base_digits()
+            .chars()
+            .any(|y| x.to_ascii_lowercase() == y)
```

Comment:
This function is kind of wasteful. Can you at least move the case conversion outside the closure?